p2p/host/basic/basic_host.go

```diff
@@ -700,9 +732,39 @@ func (h *BasicHost) AllAddrs() []ma.Multiaddr {
 		natMappings = h.natmgr.NAT().Mappings()
 	}
 
-	finalAddrs := listenAddrs
-	if len(natMappings) > 0 {
+	var finalAddrs []ma.Multiaddr
+
+	for _, a := range listenAddrs {
```

Comment:
Thinking through this, we could also just call `ListenAddresses` instead of `InterfaceListenAddresses` and manually replace all instances of the "zero" address with `h.localIPv4Addr`. That way, if the user _explicitly_ listens on a specific address (not `0.0.0.0`), we preserve it instead of filtering it out.

Reply:
@Stebalien Going through the code for those functions, I think `InterfaceListenAddresses` ALSO returns the addresses the user explicitly listens on. I wrote a small test to confirm this:

```
func TestSomething(t *testing.T) {
	ctx := context.Background()
	host1 := New(swarmt.GenSwarm(t, ctx, swarmt.OptDialOnly))

	require.NoError(t,host1.Network().Listen(ma.StringCast("/ip4/192.168.1.11/tcp/0")))
	fmt.Println("Host 1 All", host1.Addrs())
	fmt.Println("Host 1 Listen", host1.Network().ListenAddresses())

	host2 :=  New(swarmt.GenSwarm(t, ctx, swarmt.OptDialOnly))
	require.NoError(t, host2.Network().Listen(ma.StringCast("/ip4/0.0.0.0/tcp/0")))
	fmt.Println("Host 2 All", host2.Addrs())
	fmt.Println("Host 2 Listen", host2.Network().ListenAddresses())
}
```

```
=== RUN   TestSomething
Host 1 All [/ip4/192.168.1.11/tcp/52375]
Host 1 Listen [/ip4/192.168.1.11/tcp/52375]
Host 2 All [/ip4/127.0.0.1/tcp/52376 /ip4/192.168.1.11/tcp/52376]
Host 2 Listen [/ip4/0.0.0.0/tcp/52376]
```

However, if the "explicit address" is a private address that is NOT the default route, it will get filtered. Using `ListenAddresses` address here will ALSO not solve the problem.  Do we need to make any change here ?

Reply:
> However, if the "explicit address" is a private address that is NOT the default route, it will get filtered.

That's my concern. I'm saying that we _shouldn't_ filter. At the moment, we're:

1. Resolving `/ip4/0.0.0.0/tcp/4001` to `/ip4/192.168.0.99/tcp/4001`, `/ip4/10.0.1.20/tcp/4001`, etc. in the swarm.
2. Retroactively filtering all non-localhost interface addresses _except_ the default source address.

Instead, we could just:

1. Call `ListenAddresses` to get the _exact_ listen addresses that the user used.
2. Resolve any instances of the "all" address (`0.0.0.0`) to the default source address.

This will:

1. Avoid doing a bunch of unnecessary work to _undo_ the work we're doing in `InterfaceListenAddresses`.
2. Ensure that, if the user _explicitly_ listens on an interface that _isn't_ the default one, that address will still be properly advertised.

Reply:
@Stebalien This really makes sense. Have made the change.